src/main/scala/firrtl/passes/memlib/ReplaceMemMacros.scala

```diff
@@ -24,8 +24,9 @@ object ReplaceMemMacros {
   * This will not generate wmask ports if not needed.
   * Creates the minimum # of black boxes needed by the design.
   */
-class ReplaceMemMacros(writer: ConfWriter) extends Transform with DependencyAPIMigration {
-
+class ReplaceMemMacros extends Transform with DependencyAPIMigration {
+  val defAnnotatedMemories: collection.mutable.ListBuffer[DefAnnotatedMemory] =
+    collection.mutable.ListBuffer[DefAnnotatedMemory]()
```

Comment:
This shouldn't be public, and also must be cleared each time the transform is run. Ideally we wouldn't have a class field for this information (arguments are better), but that obviously would be a much more intrusive change.

Reply:
> This shouldn't be public, and also must be cleared each time the transform is run.

Are you telling me that if my Transform is a class, only a single instance is created and could be used for multiple calls to `execute`? If that is the case, shouldn't all transforms be `object` instead of `class`?

Reply:
Yes, the dependency manager will only construct one instance of a class if it needs to run multiple times. Maybe this wasn't the best idea, but the intent was to reuse existing transforms to solve lowerings due to invalidations. This can be a source of bugs (like this) where there is state in the transform.

Reply:
> Maybe this wasn't the best idea, but the intent was to reuse existing transforms to solve lowerings due to invalidations.

I see. I guess that would mean that eventually we just migrate all transforms to be `object` and then maybe we can deprecate `Dependency[..]` and leave `Dependency(...)` as the only option.

(_Note_:  sorry for hijacking this thread)

Reply:
(2) may also require defining a critical region to serialize accesses to the mutable internals. I think there can be bugs when working in a parallel environment using the same stage. This can happen in tests.

@ekiwi: That seems very reasonable, though a long, painful deprecation road. :)

Reply:
> (2) may also require defining a critical region to serialize accesses to the mutable internals. I think there can be bugs when working in a parallel environment using the same stage. This can happen in tests.

Got it, I will directly refactor it ;)

> (Note: sorry for hijacking this thread)

Haha, no worries, I think this did help me understand what I should do for this PR.